File: src/test/java/kitchenpos/stringcalculator/InputParser.java
package kitchenpos.stringcalculator;

import java.util.regex.Matcher;
import java.util.regex.Pattern;

public class InputParser {
    private static final String DEFAULT_DELIMITERS = ",|:";
    private static final Pattern CUSTOM_DELIMITER_PATTERN = Pattern.compile("//(.)\n(.*)");
    private static final int DELIMITER_GROUP = 1;
    private static final int NUMBERS_GROUP = 2;

    private InputParser() {
    }

    public static String[] parse(String input) {
        Matcher matcher = CUSTOM_DELIMITER_PATTERN.matcher(input);
        if (matcher.find()) {
            String customDelimiter = matcher.group(DELIMITER_GROUP);
            String numbers = matcher.group(NUMBERS_GROUP);
            return numbers.split(customDelimiter);
        }

        return input.split(DEFAULT_DELIMITERS);
    }
}

File: src/test/java/StringCalculator/InputParser.kt
package StringCalculator

object InputParser {

    private const val DEFAULT_DELIMITERS = "[,:]"
    private val CUSTOM_DELIMITER_PATTERN = Regex("//(.)\n(.*)")

    fun parse(inputText: String?): List<String> {
        if (inputText.isNullOrEmpty()) {
            return emptyList()
        }

        val (delimiter, numbersString) = extractDelimiterAndNumbersString(inputText)
        return numbersString.split(delimiter)
    }

    private fun extractDelimiterAndNumbersString(text: String): Pair<Regex, String> {
        val matchResult = CUSTOM_DELIMITER_PATTERN.find(text)

        return if (matchResult != null) {
            val customDelimiter = matchResult.groupValues[1]
            val numbersString = matchResult.groupValues[2]
            val escapedDelimiter = Regex.escape(customDelimiter)
            Regex(escapedDelimiter) to numbersString
        } else {
            Regex(DEFAULT_DELIMITERS) to text
        }
    }
}